### PR TITLE
Added Sample HIU to test consent related workflows.

### PR DESCRIPTION
**Pull request by atulai-sg**
- Fixed consent workflows. (Still need to test consent-status and fetch-consents)
- Added Sample HIU springboot application to test consent use cases.

### CODE REVIEW

File: src/main/java/com/nha/abdm/wrapper/hip/hrp/database/mongo/services/RequestLogService.java [hunk 1]
```diff
@@ -356,28 +355,32 @@ public void persistConsentInitRequest(
...
+  //    Map<String, Object> map = new HashMap<>();
+  //    map.put(FieldIdentifiers.CONSENT_REQUEST_ID, consentRequestId);
+  //    Update update = new Update();
+  //    update.set(FieldIdentifiers.RESPONSE_DETAILS, map);
+  //    update.set(FieldIdentifiers.STATUS, RequestStatus.CONSENT_ON_INIT_RESPONSE_RECEIVED);
+  //    mongoTemplate.updateFirst(query, update, RequestLog.class);
+  //  }
 
-  public void updateConsentResponse(String requestId, ConsentStatus consentDetails)
+  public <T> void updateConsentResponse(
```
Comment (chandrashekar-s) [inline src/main/java/com/nha/abdm/wrapper/hip/hrp/database/mongo/services/RequestLogService.java:368]: Good to have: Is it a good idea to have some indicator in the `RequestLog` table saying what is the request about? like `CONSENT_CREATION`, `CARE_CONTEXT_LINKING`, `DATA_TRANSFER_FETCH` etc?

**Review by chandrashekar-s — COMMENTED** [review]
Thanks @atulai-sg for the changes. There are few minor comments/suggestions, please have a look. Also, in general the `RequestLogService` has functionalities of multiple things, so I am thinking if there are ways of simplifying this. WDYT?

File: sample-hiu/specs/hiu-facade.yaml [hunk 2]
```diff
@@ -0,0 +1,288 @@
+openapi: 3.0.3
```
[the remark this answers is not in the document]
Comment (atulai-sg) [reply sample-hiu/specs/hiu-facade.yaml:1]: This is nice, created an issue for this: https://github.com/NHA-ABDM/ABDM-wrapper/issues/75

File: sample-hiu/specs/hiu-facade.yaml [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+              properties:
+                requestId:
+                  type: string
+                timestamp:
+                  type: string
+                consent:
+                  $ref: '#/components/schemas/ConsentRequest'
+      responses:
+        '200':
+          description: OK
```
[the remark this answers is not in the document]
Comment (atulai-sg) [reply sample-hiu/specs/hiu-facade.yaml:32]: ack

File: src/main/java/com/nha/abdm/wrapper/hip/hrp/database/mongo/MongoConfig.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+import org.springframework.data.mongodb.config.AbstractMongoClientConfiguration;
+import org.springframework.transaction.annotation.EnableTransactionManagement;
+
+/**
+ * This class can be used to provide Transactional capabilities to our database operations if we
+ * have set up mongo database with replica sets. We can then annotate methods with Transactional
+ * annotation, and then those methods should honor the transactional properties.
+ */
+@Configuration
+@EnableTransactionManagement
```
[the remark this answers is not in the document]
Comment (atulai-sg) [reply src/main/java/com/nha/abdm/wrapper/hip/hrp/database/mongo/MongoConfig.java:19]: ack

File: src/main/java/com/nha/abdm/wrapper/hip/hrp/database/mongo/services/RequestLogService.java [hunk 5]
```diff
@@ -356,28 +355,32 @@ public void persistConsentInitRequest(
...
-  public void updateConsentResponse(String requestId, String consentRequestId) {
-    Query query = new Query(Criteria.where(FieldIdentifiers.GATEWAY_REQUEST_ID).is(requestId));
-    Map<String, Object> map = new HashMap<>();
-    map.put(FieldIdentifiers.CONSENT_REQUEST_ID, consentRequestId);
-    Update update = new Update();
-    update.set(FieldIdentifiers.RESPONSE_DETAILS, map);
-    update.set(FieldIdentifiers.STATUS, RequestStatus.CONSENT_ON_INIT_RESPONSE_RECEIVED);
-    mongoTemplate.updateFirst(query, update, RequestLog.class);
-  }
+  //  public void updateConsentResponse(String requestId, String consentRequestId) {
```
[the remark this answers is not in the document]
Comment (atulai-sg) [reply src/main/java/com/nha/abdm/wrapper/hip/hrp/database/mongo/services/RequestLogService.java:358]: cleaned it, thanks. left by mistake

File: src/main/java/com/nha/abdm/wrapper/hiu/hrp/consent/GatewayCallbackService.java [hunk 6]
```diff
@@ -5,91 +5,109 @@
...
+    } else if (Objects.nonNull(onInitRequest)
+        && Objects.nonNull(onInitRequest.getResp())
+        && Objects.nonNull(onInitRequest.getError())) {
       requestLogService.updateError(
-          onInitRequest.getRequestId(),
+          onInitRequest.getResp().getRequestId(),
           onInitRequest.getError().getMessage(),
           RequestStatus.CONSENT_ON_INIT_ERROR);
+    } else if (Objects.nonNull(onInitRequest) && Objects.nonNull(onInitRequest.getResp())) {
+      requestLogService.updateError(
```
[the remark this answers is not in the document]
Comment (atulai-sg) [reply src/main/java/com/nha/abdm/wrapper/hiu/hrp/consent/GatewayCallbackService.java:71]: ack

File: src/main/resources/application.properties [hunk 7]
```diff
@@ -38,6 +38,9 @@ getPatientPath=http://localhost:8081/v1/patients
 
 logging.level.com.nha.abdm.wrapper=DEBUG
 logging.level.org.springframework.web=DEBUG
+logging.level.org.springframework.web.servlet.mvc.method.annotation.HttpEntityMethodProcessor=TRACE
```
[the remark this answers is not in the document]
Comment (atulai-sg) [reply src/main/resources/application.properties:41]: These provide the request details, if we don't do this then request details get truncated. Again, this does not end up generating verbose logs. It generated trace level for just that package which is only one line for one request.

File: src/main/java/com/nha/abdm/wrapper/hiu/hrp/consent/GatewayCallbackService.java [hunk 8]
```diff
@@ -5,91 +5,109 @@
...
-      requestLogService.updateError(
-          hiuConsentOnStatusRequest.getRequestId(),
-          hiuConsentOnStatusRequest.getError().getMessage(),
-          RequestStatus.CONSENT_ON_STATUS_ERROR);
+          gatewayRequestId,
+          FieldIdentifiers.CONSENT_ON_STATUS_RESPONSE,
+          RequestStatus.CONSENT_ON_STATUS_RESPONSE_RECEIVED,
+          hiuConsentOnStatusRequest.getConsentRequest());
+    } else {
+      // There is no way to track the gateway request id since gateway sent empty request. So we
```
[the remark this answers is not in the document]
Comment (atulai-sg) [reply src/main/java/com/nha/abdm/wrapper/hiu/hrp/consent/GatewayCallbackService.java:92]: ack

File: src/main/java/com/nha/abdm/wrapper/hip/hrp/database/mongo/services/RequestLogService.java (hunk 1 above)
Comment (atulai-sg) [reply src/main/java/com/nha/abdm/wrapper/hip/hrp/database/mongo/services/RequestLogService.java:368]: we can have that to enable hinting or if there is a concrete requirement like what we will do with this hint or metadata?

File: src/main/java/com/nha/abdm/wrapper/hiu/hrp/consent/HIUConsentService.java [hunk 9]
```diff
@@ -110,56 +112,85 @@ public ConsentStatusResponse consentRequestStatus(String clientRequestId)
           "Client request not found in database: " + clientRequestId);
     }
     try {
-      // Check whether we have got consentRequestId as part of consent on init call from gateway.
-      // If not, then send the request status as is.
-      if (Objects.isNull(requestLog.getResponseDetails())
-          || Objects.isNull(
-              requestLog.getResponseDetails().get(FieldIdentifiers.RESPONSE_DETAILS))) {
+      // Check whether we have got consent response as part of 'consent hiu-notify'.
```
[the remark this answers is not in the document]
Comment (atulai-sg) [reply src/main/java/com/nha/abdm/wrapper/hiu/hrp/consent/HIUConsentService.java:115]: ack

File: src/main/java/com/nha/abdm/wrapper/hip/hrp/database/mongo/services/RequestLogService.java (hunk 1 above)
Comment (atulai-sg) [reply src/main/java/com/nha/abdm/wrapper/hip/hrp/database/mongo/services/RequestLogService.java:361]: Have created an issue to simplify this class: https://github.com/NHA-ABDM/ABDM-wrapper/issues/76

**Pull request merged**